### DISCUSSION TITLE
Issuance of credentials via Bulk Upload

### DISCUSSION THREAD

**Discussion by rhwarrier** [Sunbird-RC/community#398] [Ideas & Enhancements]
Does RC support bulk issuance of credentials via file upload ? 
There are some references to this in some of the tickets (https://github.com/Sunbird-RC/community/issues/203) just wanted to check if bulk issuance has already been built?

(If not, we'll have a new issue created for it - this is something that needs to be prioritised for the Education Registry use case.

**Comment by snehal0904** [comment]
We built a bulk API to issue the credentials in bulk. This is not directly merged to RC we built middleware/BFF (backend for frontend) for it and this BFF service communicates with RC.

**What does the bulk API do**
1. Accept one or more records to issue credentials for
2. Lookup subject details based on the subject's fields from a defined RC schema
3. Lookup issuer details based on the issuer's fields from a defined RC schema
4. Get VC schema
5. Loop through the records and call the VC issue API from the new credential microservice for each record


To make this bulk API generic, we need to support the below configurations. This configuration.json can mount in the docker volume and will get configured as per the credential requirements. This configuration JSON will have the following configurations,

1. Which is the RC schema for subject lookup
2. Which are the lookup fields for subject data
3. Which is the RC schema for issuer lookup
4. Which are the lookup fields for issuer data

Example configuration JSON file - 
```{
  "subject":{
       "schema": "student",
       "fields":[
         "name", "enrolledOn","dateOfBirth","rollNo"
       ]
  },
  "issuer":{
    "schema": "school",
    "fields":[
      "name","udise","address", "type"
    ]
  }
}
```

**Assumptions**
1. The fields to be included in the VC payload need to be present in the RC schemas for the issuer & subject
2. BFF will check record already exist for the issuer, if it exists it will pick the existing DID and send VC credentials to API

Proposed request body for the bulk API
```
POST /api/v1/bulkupload/:vcschematype
```
**Example Payload -** 
```
{
  "issuerDetail": {
    "issuerField1": "value",
     "issuerField2": "value"
  },
  "vcData": {
    "issuanceDate": "2023-02-06T11:56:27.259Z",
    "expirationDate": "2023-02-06T11:56:27.259Z"
  },
  "credentialSubjectCommon": {
     "subjectCommonField1": "value",
     "subjectCommonField2": "value"
  },
  "credentialSubject": [
    {
      "subjectField1": "value",
      "subjectField2": "value",
      "subjectField3": "value"
    },
  {
      "subjectField1": "value",
      "subjectField2": "value",
      "subjectField3": "value"
    }
  ]
}

```
cc @coolbung @parthlawate @surendrasinghs
